### PR TITLE
feat: add property accessor to PagerRenderer

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Add property accessor to PagerRenderer.
To generate Pager HTML that is the same as CI3's.
The template would be like this:
```php
    <?php if ($pager->hasPreviousPage()) : ?>
    <a href="<?= $pager->getFirst() ?>" data-ci-pagination-page="<?= $pager->first ?>" rel="start">&laquo;First</a>
    <a href="<?= $pager->getPreviousPage() ?>" data-ci-pagination-page="<?= $pager->current - 1 ?>" rel="prev">&lt;</a>
    <?php endif ?>
```

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Pager/PagerRenderer.php [hunk 1]
```diff
@@ -97,6 +99,25 @@ public function __construct(array $details)
...
+	 * @return mixed
+	 */
+	public function __get(string $key)
+	{
+		if (property_exists($this, $key))
+		{
+			return $this->$key;
+		}
+
+		throw new InvalidArgumentException('No such a property: ' . $key);
```
Comment (paulbalandan) [inline system/Pager/PagerRenderer.php:116]:
```suggestion
		throw new InvalidArgumentException('No such property: ' . $key);
```

**Review by paulbalandan — COMMENTED** [review]
Please rebase.

File: tests/system/Pager/PagerRendererTest.php [hunk 2]
```diff
@@ -560,4 +561,39 @@ public function testGetNextPageWithSegmentHigherThanZero()
...
+
+		$this->assertEquals(1, $pager->first);
+		$this->assertEquals(3, $pager->current);
+		$this->assertEquals(10, $pager->last);
+	}
+
+	public function testPropertyAccessorGetPropertyThatDoesNotExists()
+	{
+		$this->expectException(InvalidArgumentException::class);
+		$this->expectExceptionMessage('No such a property: properyNeverEverExists');
```
Comment (MGatner) [inline tests/system/Pager/PagerRendererTest.php:585]:
```suggestion
		$this->expectExceptionMessage('No such property: properyNeverEverExists');
```
Need to update the test case too!

**Review by MGatner — CHANGES_REQUESTED** [review]

**Comment by iRedds** [comment]
Maybe change all properties to public?

**Comment by kenjis** [comment]
@iRedds I don't need public properties. I just need to get (read) them, not need to write them.

**Comment by iRedds** [comment]
@kenjis Who cares. Properties are defined in the constructor.

File: tests/system/Pager/PagerRendererTest.php (hunk 2 above)
Comment (kenjis) [reply tests/system/Pager/PagerRendererTest.php]: @MGatner Thank you. I fixed.

**Comment by kenjis** [comment]
@iRedds I don't need public properties. Do you need public properties?

**Review by paulbalandan — APPROVED** [review]

**Comment by iRedds** [comment]
If you don't need public properties, don't change their values.
I don't see any point in this pager as part of the framework. I understand that you are doing them read-on.

But why not just make them public? They only work with a template anyway.

**Comment by MGatner** [comment]
Public properties should be avoided unless the class is intentionally a mutable one (like Entity). I would like less public properties in the framework overall, not more. An accessory is the correct way to handle this.

**Comment by michalsn** [comment]
Since we are working with these properties inside the class, we shouldn't allow them to be set from outside the class as this can lead to problems.

**Review by michalsn — APPROVED** [review]

**Comment by iRedds** [comment]
It may be better to use getters in this case than the magic method?

**Comment by MGatner** [comment]
Getters would be the standard way but I believe the point of the PR was to allow backwards-compatibility with the "CI3 way". I don't have a particular issue with the magic method as long as it doesn't mess up the static analysis.

**Comment by kenjis** [comment]
@iRedds 
I just implemented getters with the magic method.
Do you mean having 6 getter methods like `getCurrent()` is better than having `__get()`?

**Comment by iRedds** [comment]
@kenjis In this case, any property added to the class in the future, but not intended for public use, will be automatically accessed using the magic method.
I think this is not entirely correct.

**Comment by kenjis** [comment]
@iRedds 
When a property will be added to the class in the future, but not intended for public use, 
we can restrict accessing the property.

**Comment by MGatner** [comment]
In general I think using getter methods is better. @kenjis I would like to hear how much of the goal of this PR is supporting the CodeIgniter 3 direct access method versus just getting at the data.

**Comment by michalsn** [comment]
The only downside is that "natural" getters for these properties are in most cases already taken, so we would have to come up with something more "complicated". And this is probably the only reason in favor of using the magic method for me.

Having something like `$pager->getPropertyPrevious()` would be more elegant than writing `$pager->current - 1`. However, I'm not sure about the naming convention.

**Comment by kenjis** [comment]
@MGatner What do you mean by `supporting the CodeIgniter 3 direct access method`?

I thought PagerRenderer has data only for pagination rendering, so it is more flexible that the view can access all the properties inf the PagerRenderer.

But in my use case, I don't need all the properties.
I'm migrating CI3 app to CI4, and `default_full.php` has became like this:
```php
<?php

/**
 * @var \CodeIgniter\Pager\PagerRenderer $pager
 */

$pager->setSurroundCount(2);
?>

<p class="pagination">
<?php if ($pager->hasPreviousPage()) : ?>
    <?php if ($pager->current !== 2) : ?>
    <a href="<?= $pager->getFirst() ?>" data-ci-pagination-page="<?= $pager->first ?> rel="start">&laquo;<?= lang('Pager.first') ?></a>
    <?php endif ?>

    <a href="<?= $pager->getPreviousPage() ?>" data-ci-pagination-page="<?= $pager->current - 1 ?>" rel="prev">&lt;</a>
<?php endif ?>

<?php foreach ($pager->links() as $link) : ?>
<?php if ($link['active']) : ?>
    <strong><?= $link['title'] ?></strong>
<?php else : ?>
    <a href="<?= $link['uri'] ?>" data-ci-pagination-page="<?= $link['title'] ?>">
        <?= $link['title'] ?>
    </a>
<?php endif ?>
<?php endforeach ?>

<?php if ($pager->hasNextPage()) : ?>
    <a href="<?= $pager->getNextPage() ?>" data-ci-pagination-page="<?= $pager->current + 1 ?>" rel="next">&gt;</a>

    <?php if ($pager->current !== ($pager->last - 1)) : ?>
    <a href="<?= $pager->getLast() ?>" data-ci-pagination-page="<?= $pager->last ?>"><?= lang('Pager.last') ?>&raquo;</a>
    <?php endif ?>
<?php endif ?>
</p>
```

**Comment by MGatner** [comment]
This class is very confusing because "first", "last", "current" etc are used interchangeably for page numbers and for URIs. I think we might do better to differentiate those, so instead of having `$pager->first` (number) and `$pager->getFirst()` (URL) we have some clearer accessors.

**Comment by kenjis** [comment]
How about these?
```
getFirstPageNumber()
getCurrentPageNumber()
getLastPageNumber()
```

**Comment by kenjis** [comment]
I take down this PR.
To make all properties readable is too open. 
https://github.com/codeigniter4/CodeIgniter4/blob/49520c7862ab7ce42bb7da54df89a04e33d49fa1/system/Pager/PagerRenderer.php#L45

I sent another PR: #4188

**Pull request closed without merging**